From eaa02103a7337470f06a857bb97fa1e67f448b4b Mon Sep 17 00:00:00 2001
From: Andrew Wang <wang@apache.org>
Date: Wed, 19 Nov 2014 17:53:00 -0800
Subject: [PATCH 245/596] HDFS-7409. Allow dead nodes to finish
 decommissioning if all files are fully replicated.

(cherry picked from commit 765aecb4e127ebaf6c7b060a8b5cd40c6428e296)
(cherry picked from commit 988682ac2957da0f34e9fe524d71c5931365fe95)
(cherry picked from commit 332f353bcdaf911d1a4d3f5b8cc38e8d2bcef1f8)
(cherry picked from commit 44cdd3b96e9091bc79fac5485f1d2ba6108ceabd)
---
 .../hdfs/server/blockmanagement/BlockManager.java  |   18 +++++++++++++-----
 .../server/namenode/TestDecommissioningStatus.java |   13 ++++++++++---
 2 files changed, 23 insertions(+), 8 deletions(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/BlockManager.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/BlockManager.java
index b2c7391..675c6f1 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/BlockManager.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/BlockManager.java
@@ -3275,11 +3275,19 @@ boolean isReplicationInProgress(DatanodeDescriptor srcNode) {
     }
 
     if (!status && !srcNode.isAlive) {
-      LOG.warn("srcNode " + srcNode + " is dead " +
-          "when decommission is in progress. Continue to mark " +
-          "it as decommission in progress. In that way, when it rejoins the " +
-          "cluster it can continue the decommission process.");
-      status = true;
+      updateState();
+      if (pendingReplicationBlocksCount == 0 &&
+          underReplicatedBlocksCount == 0) {
+        LOG.info("srcNode {} is dead and there are no under-replicated" +
+            " blocks or blocks pending replication. Marking as " +
+            "decommissioned.");
+      } else {
+        LOG.warn("srcNode " + srcNode + " is dead " +
+            "while decommission is in progress. Continuing to mark " +
+            "it as decommission in progress so when it rejoins the " +
+            "cluster it can continue the decommission process.");
+        status = true;
+      }
     }
 
     srcNode.decommissioningStatus.set(underReplicatedBlocks,
diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/TestDecommissioningStatus.java b/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/TestDecommissioningStatus.java
index c6312c0..a2b7b87 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/TestDecommissioningStatus.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/TestDecommissioningStatus.java
@@ -18,6 +18,7 @@
 package org.apache.hadoop.hdfs.server.namenode;
 
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertTrue;
 
 import java.io.IOException;
@@ -359,8 +360,15 @@ public void testDecommissionStatusAfterDNRestart()
     BlockManagerTestUtil.checkDecommissionState(dm, dead.get(0));
 
     // Verify that the DN remains in DECOMMISSION_INPROGRESS state.
-    assertTrue("the node is in decommissioned state ",
-        !dead.get(0).isDecommissioned());
+    assertTrue("the node should be DECOMMISSION_IN_PROGRESSS",
+        dead.get(0).isDecommissionInProgress());
+
+    // Delete the under-replicated file, which should let the 
+    // DECOMMISSION_IN_PROGRESS node become DECOMMISSIONED
+    cleanupFile(fileSys, f);
+    BlockManagerTestUtil.checkDecommissionState(dm, dead.get(0));
+    assertTrue("the node should be decommissioned",
+        dead.get(0).isDecommissioned());
 
     // Add the node back
     cluster.restartDataNode(dataNodeProperties, true);
@@ -371,7 +379,6 @@ public void testDecommissionStatusAfterDNRestart()
     // make them available again.
     writeConfigFile(localFileSys, excludeFile, null);
     dm.refreshNodes(conf);
-    cleanupFile(fileSys, f);
   }
 
   /**
-- 
1.7.9.5

